### PR TITLE
Optional method name in task deferred

### PR DESCRIPTION
**Pull request by malthe**
Currently, a `TaskDeferred` takes a `method_name` parameter such as "execute_complete" and this method must then exist on the operator.
```python
class MyOperator(BaseOperator):
    def execute_complete(self, context, event=None):
        """Callback for when the trigger fires."""
```

But in some cases, it makes sense to simply restart the operator – essentially calling its "execute" method. There is however an inconvenience or compatibility issue here in that the resume method receives an "event" keyword argument with the trigger event payload – and "execute" takes no such argument.

To appreciate why this is a problem, consider the problem of having to delay the execution of an existing operator component, for example deferring its execution for 10 minutes. We would like to avoid having to extend the implementation of this operator in order to accomplish this.

In #17576, support for `pre_execute` was added – a hook that runs before the operator's `execute` method. The change proposal presented here builds upon this functionality to allow an implementation such as:
```python
class PreExecuteDefer:
    def __init__(self, timedelta):
        self._timedelta = timedelta

    def __call__(self, context):
        moment = context["data_interval_end"] + self._timedelta

        # We must verify the trigger condition since upon re-entry, this will be called again.
        if moment < timezone.utcnow():
            return

        trigger = DateTimeTrigger(moment=moment)
        raise TaskDeferred(trigger=trigger, kwargs={})
```
This can be used as the `pre_execute` hook to the effect of delaying the execution.

### CODE REVIEW

**Comment by kaxil** [comment]
cc @andrewgodwin @jarfgit

**Comment by andrewgodwin** [comment]
I think this looks reasonable - I had not considered the potential interaction with `pre_execute` like a delay, but it makes sense.

File: airflow/jobs/scheduler_job.py [hunk 1]
```diff
@@ -533,7 +533,7 @@ def _process_executor_events(self, session: Session = None) -> int:
         return len(event_buffer)
 
     def _execute(self) -> None:
-        self.log.info("Starting the scheduler")
+        self.log.info("Starting the scheduler: %r", self.executor)
```
Comment (ashb) [inline airflow/jobs/scheduler_job.py]: Please avoid making changes to unrelated files in PRs.

File: airflow/exceptions.py [hunk 2]
```diff
@@ -256,12 +278,21 @@ def __init__(
...
         self.timeout = timeout
+
         # Check timeout type at runtime
         if self.timeout is not None and not hasattr(self.timeout, "total_seconds"):
             raise ValueError("Timeout value must be a timedelta")
 
+        # Check keyword arguments
+        if kwargs:
+            if method_name is None:
+                raise ValueError("Keyword arguments not allowed when method is not specified")
```
Comment (ashb) [inline airflow/exceptions.py:286]: Why not  just default `method_name: str = 'execute'` instead?

File: airflow/exceptions.py [hunk 3]
```diff
@@ -256,12 +278,21 @@ def __init__(
...
+            else:
+                if "event" in kwargs:
+                    raise ValueError("Keyword argument must not contain reserved name 'event'")
```
Comment (ashb) [inline airflow/exceptions.py:289]: Shouldn't this  apply no mater what the method_name is?

**Review by ashb — CHANGES_REQUESTED** [review]
> There is however an inconvenience or compatibility issue here in that the resume method receives an "event" keyword argument with the trigger event payload – and "execute" takes no such argument.

Can't you do this:

```python

class MyOperator(BaseOperator):
    def execute(self, context, event=None):
        if not event:
            self.defer(method_name='execute')
```


At any rate, I don't quite understand why this PR is needed, nor how it relates to the problem of pre_execute/post_execute -- from my reading of the code once a task is deferred and resumed then the post_execute code won't get run, and this PR doesn't change that.



~Ohhhh, by deferring with an empty method_name it goes back to the normal pre+execute+post flow.~

~That feels a little bit too much like magic for me, and does nothing to address the fact that `post_execute` won't work for async operators that _do_ resume to a different method.~

Nope, my first reading was right. I don't understand why this change is needed.

File: airflow/exceptions.py (hunk 3 above)
Comment (malthe) [reply airflow/exceptions.py:289]: If `method_name` is unset then `kwargs` are disallowed period – so we don't bother checking them.

File: airflow/exceptions.py (hunk 2 above)
Comment (malthe) [reply airflow/exceptions.py:286]: So the only reason is to circumvent the behavior of adding "event" to the `kwargs` which ruins the logic for an operator that just implements the standard `execute` signature.

It makes it possible to use `pre_execute` as a general mechanism (or workaround) for adding logic before an operator runs – that's at least my example use-case here, delaying any odd operator without making a change.

Comment (andrewgodwin) [reply airflow/exceptions.py:289]: Question: is the only saving from this not putting `event=None` in your `execute()` method signature? Is that worth making a whole bunch of the API slightly more complex?

Comment (malthe) [reply airflow/exceptions.py:286]: The problem is that it's often times not my `execute()` method – lots of people use built-in operators or write their own ones that don't know about or want to know about deferred tasks.

Comment (andrewgodwin) [reply airflow/exceptions.py:286]: Recorrecting my thoughts here - it should be fine to have custom kwargs for `execute`, since it's something you can always override in an operator.

Comment (andrewgodwin) [reply airflow/exceptions.py:286]: > don't know about or want to know about deferred tasks

Wait, so you're injecting deferral into operators via the pre_execute hook that otherwise may not expect it? I would be wary of doing that, there might be side-effects from operators that do too much in their own init or pre-execute code...

Comment (ashb) [reply airflow/exceptions.py:286]: > It makes it possible to use pre_execute as a general mechanism (or workaround) for adding logic before an operator runs – that's at least my example use-case here, delaying any odd operator without making a change.

This sounds like a bit of an anti-pattern to me -- TimeDeltaSensor (+ an Async version) already exists, or change the schedule of the whole DAG to be N minutes later.

Comment (malthe) [reply airflow/exceptions.py:286]: On not allowing `kwargs` when `method_name` is unset, I just thought it was a little cleaner. I'm not sure I understand the part about custom kwargs for `execute` – how does that look/work?

Is this an anti-pattern? So I don't think so and there is some more discussion on this in #17576. For example, if I want to postpone a particular task (not the whole DAG, just a task) so that it runs no earlier than 7:30am then to me, that is a scheduling policy on the task, it's not helpful to have another "delayer" task in front of it because ultimately, that just contributes to noise.

I'm migrating some workflows from a different system where there is task-level scheduling that allows such configurations and it really works quite well. There is less clutter when each task has a little bit more "scheduling power". But of course, that's just a point of view.

Comment (malthe) [reply airflow/exceptions.py:286]: @andrewgodwin note that the operator should not set a `pre_execute` hook on itself, that would be weird I think due to the semantics around `pre_execute` – it runs sequentially just before `execute` so you could just put that logic into the `execute` method.

The idea with `pre_execute` is that a user can have some custom logic before the operator executes. They could subclass and come up with some name for the class, but this is a simple way to change the execution flow on an ad-hoc basis.

**Comment by malthe** [comment]
@andrewgodwin a different take on this issue could be to add the following method to `BaseOperator`:
```python
def execute_complete(self, context, event=None):
    self.execute(context)
```
(With an appropriate docstring of course and some added notes in "deferring.rst".
